File: Sources/RMBTQoSTCPTest.m
/*
 * Copyright 2017 appscape gmbh
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "RMBTQoSTCPTest.h"
#import <CocoaAsyncSocket/GCDAsyncSocket.h>
#import "RMBT-Swift.h"

@interface RMBTQoSTCPTest()<GCDAsyncSocketDelegate> {
    dispatch_semaphore_t _doneSem;
    NSString *_response;
}
@end

@implementation RMBTQoSTCPTest
-(instancetype)initWithParams:(NSDictionary *)params {
    if (self = [super initWithParams:params]) {
        if (self.outPort > 0 && self.inPort == 0) {
            self.direction = RMBTQoSIPTestDirectionOut;
        } else if (self.inPort > 0 && self.outPort == 0) {
            self.direction = RMBTQoSIPTestDirectionIn;
        }
    }
    return self;
}

- (NSDictionary*)result {
    NSMutableDictionary *result = [@{@"tcp_objective_timeout": @(self.timeoutNanos)} mutableCopy];
    BOOL outgoing = (self.direction == RMBTQoSIPTestDirectionOut);

    if (outgoing) {
        [result addEntriesFromDictionary:@{
           @"tcp_result_out_response": RMBTValueOrNull(_response),
           @"tcp_objective_out_port": @(self.outPort),
           @"tcp_result_out": self.statusName
        }];
    } else {
        [result addEntriesFromDictionary:@{
           @"tcp_result_in_response": RMBTValueOrNull(_response),
           @"tcp_objective_in_port": @(self.inPort),
           @"tcp_result_in": self.statusName
       }];
    }

    return result;
}


- (void)ipMain:(BOOL)outgoing {
    NSUInteger port = outgoing ? self.outPort : self.inPort;

    NSError *error = nil;

    NSString *cmd = [NSString stringWithFormat:@"TCPTEST %@ %lu +ID%@", outgoing ? @"OUT" : @"IN",
                     (unsigned long)port,
                     self.uid];
    NSString *response1 = [self sendCommand:cmd readReply:outgoing ? YES : NO error:&error];
    if (error || (outgoing && ![response1 hasPrefix:@"OK"])) {
        [Log log:[NSString stringWithFormat:@"%@ failed: %@/%@", self, error, response1]];
        self.status = RMBTQoSTestStatusError;
        return;
    }

    dispatch_queue_t delegateQueue = dispatch_queue_create("at.rmbt.qos.tcp.delegate", DISPATCH_QUEUE_SERIAL);
    GCDAsyncSocket* socket = [[GCDAsyncSocket alloc] initWithDelegate:self delegateQueue:delegateQueue];

    if (outgoing) {
        [socket connectToHost:self.controlConnectionParams.serverAddress onPort:port withTimeout:self.timeoutSeconds error:&error];
    } else {
        [socket acceptOnPort:port error:&error];
    }

    if (error) {
        [Log log:[NSString stringWithFormat:@"%@ error connecting/binding: %@", self, error]];
        self.status = RMBTQoSTestStatusError;
        return;
    }

    _doneSem = dispatch_semaphore_create(0);
    if (dispatch_semaphore_wait(_doneSem, dispatch_time(DISPATCH_TIME_NOW, self.timeoutNanos)) != 0) {
        self.status = RMBTQoSTestStatusTimeout;
    } else if (_response && ![_response isEqualToString:@""]) {
        self.status = RMBTQoSTestStatusOk;
    }

    socket.delegate = nil;
    [socket disconnect];
}


- (void)socket:(GCDAsyncSocket *)sock didConnectToHost:(NSString *)host port:(uint16_t)port {
    NSParameterAssert(self.direction == RMBTQoSIPTestDirectionOut);
    [sock writeData:[@"PING\n" dataUsingEncoding:NSASCIIStringEncoding] withTimeout:self.timeoutSeconds tag:0];
}

- (void)socket:(GCDAsyncSocket *)sock didAcceptNewSocket:(GCDAsyncSocket *)newSocket {
    [newSocket readDataToData:[@"\n" dataUsingEncoding:NSASCIIStringEncoding] withTimeout:self.timeoutSeconds tag:0];
}

- (void)socketDidDisconnect:(GCDAsyncSocket *)sock withError:(NSError *)err {
    dispatch_semaphore_signal(_doneSem);
}

- (void)socket:(GCDAsyncSocket *)sock didWriteDataWithTag:(long)tag {
    [sock readDataToData:[@"\n" dataUsingEncoding:NSASCIIStringEncoding] withTimeout:self.timeoutSeconds tag:0];
}

- (void)socket:(GCDAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag {
    NSString *line = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
    NSParameterAssert(tag == 0);
    _response = [RMBTHelpers RMBTChomp:line];
    [sock disconnect];
}

- (NSString*)description {
    return [NSString stringWithFormat:@"RMBTQoSTCPTest (uid=%@, cg=%ld, server=%@, out_port=%ld, in_port=%ld)",
            self.uid,
            (unsigned long)self.concurrencyGroup,
            self.controlConnectionParams,
            (unsigned long)self.outPort,
            (unsigned long)self.inPort
    ];
}

@end

File: Sources/RMBTQoSTCPTest.swift
//
//  RMBTQoSTCPTest.swift
//  RMBT
//
//  Created by Sergey Glushchenko on 10.12.2021.
//  Copyright © 2021 appscape gmbh. All rights reserved.
//

import UIKit
import CocoaAsyncSocket

class RMBTQoSTCPTest: RMBTQoSIPTest {

    private var doneSem: DispatchSemaphore?
    private var response: String?
    
    override init?(with params: [String : Any]) {
        super.init(with: params)
            
        if (self.outPort > 0 && self.inPort == 0) {
            self.direction = .out
        } else if (self.inPort > 0 && self.outPort == 0) {
            self.direction = .in
        }
    }
    
    override var result: [String: Any] {
        var result: [String: Any] = ["tcp_objective_timeout": self.timeoutNanos]
        let outgoing = self.direction == .out

        if (outgoing) {
            result["tcp_result_out_response"] = RMBTValueOrNull(response)
            result["tcp_objective_out_port"] = self.outPort
            result["tcp_result_out"] = self.statusName()
        } else {
            result["tcp_result_in_response"] = RMBTValueOrNull(response)
            result["tcp_objective_in_port"] = self.inPort
            result["tcp_result_in"] = self.statusName()
        }

        return result
    }
    
    override func ipMain(_ isOutgoing: Bool) {
        let port = isOutgoing ? self.outPort : self.inPort

        let cmd = String(format: "TCPTEST %@ %lu +ID%@", isOutgoing ? "OUT" : "IN",
                         port,
                         self.uid)
        
        let response1: String
        do {
            response1 = try self.send(command: cmd, readReply: isOutgoing ? true : false)
        }
        catch let error {
            Log.logger.error("\(self) failed:\(error)")
            self.status = .error
            return
        }
        
        if (isOutgoing && !response1.hasPrefix("OK")) {
            Log.logger.error("\(self) failed:error/\(response1)")
            self.status = .error;
            return;
        }

        let delegateQueue = DispatchQueue(label: "at.rmbt.qos.tcp.delegate")
        let socket = GCDAsyncSocket(delegate: self, delegateQueue: delegateQueue)

        do {
            if (isOutgoing) {
                try socket.connect(toHost: self.controlConnectionParams?.serverAddress ?? "", onPort: UInt16(port), withTimeout: TimeInterval(self.timeoutSeconds()))
            } else {
                try socket.accept(onPort: UInt16(port))
            }
        }
        catch let error {
            Log.logger.error("\(self) error connecting/binding:\(error)")
            self.status = .error
            return
        }

        self.doneSem = DispatchSemaphore(value: 0)
        
        let result = self.doneSem?.wait(timeout: .now() + Double(self.timeoutNanos))
        if result == .timedOut {
            self.status = .timeout
        } else {
            self.status = .ok
        }
        
        socket.delegate = nil
        socket.disconnect()
    }
    
    override var description: String {
        return String(format: "RMBTQoSTCPTest (uid=%@, cg=%ld, server=%@, out_port=%ld, in_port=%ld)",
                      self.uid,
                      self.concurrencyGroup,
                      self.controlConnectionParams ?? "",
                      self.outPort,
                      self.inPort
        )
    }
}

extension RMBTQoSTCPTest: GCDAsyncSocketDelegate {
    func socket(_ sock: GCDAsyncSocket, didConnectToHost host: String, port: UInt16) {
        assert(self.direction == .out)
        sock.write("PING\n".data(using: .ascii), withTimeout: TimeInterval(self.timeoutSeconds()), tag: 0)
    }
    
    func socket(_ sock: GCDAsyncSocket, didAcceptNewSocket newSocket: GCDAsyncSocket) {
        newSocket.readData(to: "\n".data(using: .ascii), withTimeout: TimeInterval(self.timeoutSeconds()), tag: 0)
    }

    func socketDidDisconnect(_ sock: GCDAsyncSocket, withError err: Error?) {
        doneSem?.signal()
    }

    func socket(_ sock: GCDAsyncSocket, didWriteDataWithTag tag: Int) {
        sock.readData(to: "\n".data(using: .ascii), withTimeout: TimeInterval(self.timeoutSeconds()), tag: 0)
    }
    
    func socket(_ sock: GCDAsyncSocket, didRead data: Data, withTag tag: Int) {
        if let line = String(data: data, encoding: .ascii) {
            assert(tag == 0)
            response = RMBTHelpers.RMBTChomp(line)
        }
        sock.disconnect()
    }
}
